File: java/structure/linear/queue/DequeOfArray.java
package structure.linear.queue;

import java.util.Arrays;

public class DequeOfArray {
    private int[] _values = new int[64];
    private int _head = 0;
    private int _tail = 0;
    private int _capacity = 64;

    public DequeOfArray() {
        Arrays.fill(_values, 0);
    }

    public boolean isEmpty() {
        return _head == _tail;
    }

    public void enqueueHead(int value) {
        if ((_tail + 1) % _capacity == _head) {
            int[] aux = new int[_capacity + 16];
            if (_head < _tail) {
                System.arraycopy(_values, _head, aux, 0, _tail - _head);
            } else {
                System.arraycopy(_values, _head, aux, 0, _capacity - _head);
                System.arraycopy(_values, 0, aux, _capacity - _head, _tail);
            }
            Arrays.fill(aux, _capacity, _capacity + 16, 0);
            _values = aux;
            _head = 0;
            _tail = _capacity - 1;
            _capacity += 16;
        }

        _head = (0 != _head) ? _head - 1 : _capacity - 1;
        _values[_head] = value;
    }

    public void enqueueTail(int value) {
        if ((_tail + 1) % _capacity == _head) {
            int[] aux = new int[_capacity + 16];
            if (_head < _tail) {
                System.arraycopy(_values, _head, aux, 0, _tail - _head);
            } else {
                System.arraycopy(_values, _head, aux, 0, _capacity - _head);
                System.arraycopy(_values, 0, aux, _capacity - _head, _tail);
            }
            Arrays.fill(aux, _capacity, _capacity + 16, 0);
            _values = aux;
            _head = 0;
            _tail = _capacity - 1;
            _capacity += 16;
        }

        _values[_tail] = value;
        _tail = (_tail + 1) % _capacity;
    }

    public int dequeueHead() {
        if (_head != _tail) {
            int value = _values[_head];
            _head = (_head + 1) % _capacity;
            return value;
        }
        return Integer.MIN_VALUE;
    }

    public int dequeueTail() {
        if (_head != _tail) {
            _tail = (0 != _tail) ? _tail - 1 : _capacity - 1;
            return _values[_tail];
        }
        return Integer.MIN_VALUE;
    }

    public void clear() {
        Arrays.fill(_values, 0);
        _head = 0;
        _tail = 0;
    }
}

File: cpp/structure/linear/queue/DequeOfArray.cpp
#include <iostream>
#include <vector>
using namespace std;

class DequeOfArray {
public:
    DequeOfArray() : values(vector<int>(64, 0)), head(0), tail(0), capacity(64) {}

    bool isEmpty() {
        return head == tail;
    }

    void enqueueHead(int value) {
        if ((tail + 1) % capacity == head) {
            vector<int> aux(capacity + 16, 0);
            if (head < tail) {
                copy(values.begin() + head, values.begin() + tail, aux.begin());
            } else {
                copy(values.begin() + head, values.begin() + capacity, aux.begin());
                copy(values.begin(), values.begin() + tail, aux.begin() + capacity - head);
            }
            values = aux;
            head = 0;
            tail = capacity - 1;
            capacity += 16;
        }

        head = head ? head - 1 : capacity - 1;
        values[head] = value;
    }

    void enqueueTail(int value) {
        if ((tail + 1) % capacity == head) {
            vector<int> aux(capacity + 16, 0);
            if (head < tail) {
                copy(values.begin() + head, values.begin() + tail, aux.begin());
            } else {
                copy(values.begin() + head, values.begin() + capacity, aux.begin());
                copy(values.begin(), values.begin() + tail, aux.begin() + capacity - head);
            }
            values = aux;
            head = 0;
            tail = capacity - 1;
            capacity += 16;
        }

        values[tail] = value;
        tail = (tail + 1) % capacity;
    }

    int dequeueHead() {
        if (head != tail) {
            int value = values[head];
            head = (head + 1) % capacity;
            return value;
        }
        return INT_MIN;
    }

    int dequeueTail() {
        if (head != tail) {
            tail = tail ? tail - 1 : capacity - 1;
            return values[tail];
        }
        return INT_MIN;
    }

    void clear() {
        values.assign(capacity, 0);
        head = 0;
        tail = 0;
    }

private:
    vector<int> values;
    int head;
    int tail;
    int capacity;
};
